modules/command-engine/javadsl/src/main/scala/surge/javadsl/command/CommandModels.scala

```diff
@@ -2,109 +2,77 @@
 
 package surge.javadsl.command
 
-import surge.core.command.AggregateCommandModelCoreTrait
-import surge.internal.domain.CommandHandler
-import surge.internal.persistence
-import surge.javadsl._
+import surge.core.commondsl.SurgeProcessingModelCoreTrait
+import surge.internal.domain.{ SurgeContext, SurgeProcessingModel }
 import surge.javadsl.common.Context
 
 import java.util.concurrent.CompletableFuture
 import java.util.{ Optional, List => JList }
-import scala.compat.java8.FutureConverters
 import scala.compat.java8.OptionConverters._
 import scala.concurrent.{ ExecutionContext, Future }
 import scala.jdk.CollectionConverters._
+import scala.jdk.FutureConverters._
 import scala.util.Try
 
-trait AggregateCommandModel[Agg, Cmd, Evt] extends AggregateCommandModelCoreTrait[Agg, Cmd, Nothing, Evt] {
+trait AggregateCommandModel[Agg, Cmd, Evt] extends SurgeProcessingModelCoreTrait[Agg, Cmd, Evt] {
   def processCommand(aggregate: Optional[Agg], command: Cmd): JList[Evt]
   def handleEvent(aggregate: Optional[Agg], event: Evt): Optional[Agg]
 
-  final def toCore: CommandHandler[Agg, Cmd, Nothing, Evt] =
-    new CommandHandler[Agg, Cmd, Nothing, Evt] {
-      override def processCommand(ctx: persistence.Context, state: Option[Agg], cmd: Cmd): Future[CommandResult] =
-        Future.fromTry(Try(Right(AggregateCommandModel.this.processCommand(state.asJava, cmd).asScala.toSeq)))
-      override def apply(ctx: persistence.Context, state: Option[Agg], event: Evt): Option[Agg] = handleEvent(state.asJava, event).asScala
+  final def toCore: SurgeProcessingModel[Agg, Cmd, Evt] =
+    new SurgeProcessingModel[Agg, Cmd, Evt] {
+      override def handle(ctx: SurgeContext[Agg, Evt], state: Option[Agg], msg: Cmd)(implicit ec: ExecutionContext): Future[SurgeContext[Agg, Evt]] = {
+        Future.fromTry(Try(AggregateCommandModel.this.processCommand(state.asJava, msg).asScala.toSeq)).map { events =>
+          val newState = events.foldLeft(state.asJava)((s: Optional[Agg], e: Evt) => handleEvent(s, e))
+          ctx.persistEvents(events).updateState(newState.asScala).reply(state => state)
+        }
+      }
+
+      override def applyAsync(ctx: SurgeContext[Agg, Evt], state: Option[Agg], event: Evt): Future[SurgeContext[Agg, Evt]] = {
```

Comment:
This maybe wants to be revisited since the state update now happens through the `handle` method and we don't call `applyAsync` except for when applying events to the state directly (outside of the normal command flow). Ideally we can find some decent balance between a CQRS model (command handler and event handler) and generic processing model (handle this message and tell the context what you'd like to do)